File: ts/180625_bowling/bowling_180625.ts
export class Bowling {
  private rolls: number [];
  
  constructor () {
    this.rolls = [];
  }

  public roll(pins: number): void {
    this.rolls.push(pins);
  }

  public score(): number {
    let score = 0;
    let rollIndex = 0;
    for (let frame = 0; frame < 10; frame ++) {
      if (this.isStrike(rollIndex)) {
        score += (10 + this.strikeBonus(rollIndex));
        rollIndex += 1;
      } else if (this.isSpare(rollIndex)) {
        score += (10 + this.spareBonus(rollIndex));
        rollIndex += 2;
      } else {
        score += (this.rolls[rollIndex] + this.rolls[rollIndex + 1]);
        rollIndex += 2;
      }
    }
    return score;
  }

  private isStrike(rollIndex: number): boolean {
    return 10 == this.rolls[rollIndex];
  }

  private isSpare(rollIndex: number): boolean {
     return 10 == (this.rolls[rollIndex] + this.rolls[rollIndex + 1]);
  }

  private strikeBonus(rollIndex: number): number {
    return this.rolls[rollIndex + 1] + this.rolls[rollIndex + 2];
  }

  private spareBonus(rollIndex: number): number {
    return this.rolls[rollIndex + 2];
  }
}

File: java/src/main/java/bowling/Bowling180625.java
import java.util.ArrayList;

public class Bowling180625 {

    public static final int MAX_SCORE_IN_FRAME = 10;
    public static final int NORMAL_MAX_FRAME = 10;

    private ArrayList<Integer> rolls = new ArrayList<Integer>();

    public void roll(int pins) {
        this.rolls.add(pins);
    }

    public int score() {
        int score = 0;
        int rollIndex = 0;
        for (int frame = 0; frame < NORMAL_MAX_FRAME; frame ++) {
            if (isStrike(rollIndex)) {
                score += (MAX_SCORE_IN_FRAME + strikeBonus(rollIndex));
                rollIndex += 1;
            } else if (isSpare(rollIndex)) {
                score += (MAX_SCORE_IN_FRAME + spareBonus(rollIndex));
                rollIndex += 2;
            } else {
                score += (this.rolls.get(rollIndex) + this.rolls.get(rollIndex + 1));
                rollIndex += 2;
            }
        }
        return score;
    }

    private boolean isStrike(int rollIndex) {
        return MAX_SCORE_IN_FRAME == this.rolls.get(rollIndex);
    }

    private boolean isSpare(int rollIndex) {
        return MAX_SCORE_IN_FRAME == (this.rolls.get(rollIndex) + this.rolls.get(rollIndex + 1));
    }

    private int strikeBonus(int rollIndex) {
        return this.rolls.get(rollIndex + 1) + this.rolls.get(rollIndex + 2);
    }

    private int spareBonus(int rollIndex) {
        return this.rolls.get(rollIndex + 2);
    }
}
